cogs/commands/roll.py

```diff
@@ -69,152 +78,66 @@ def __init__(self, bot: Bot):
 
+        p = await get_parallelism(self.bot)
+        p.send_to_ctx_after_threaded(work, ctx, loop)
 
-class DiceParser(TextParsers):
-    nat = reg(r"[1-9]\d*") > int
 
-    constant = reg(r"\d+") > (lambda x: ValueConstant(int(x)))
-    roll = die_value << "d" & die_value > (lambda xs: ValueRoll(xs[0], xs[1]))
-    naked_roll = "d" >> die_value > (lambda x: ValueRoll(ValueConstant(1), x))
-
-    value = roll | naked_roll | constant | brackets
-    die_value = constant | brackets
-
-    op_add = value << "+" & expr > (
-        lambda xs: BinaryOperator(RollOperator.ADD, xs[0], xs[1])
-    )
-    op_sub = value << "-" & expr > (
-        lambda xs: BinaryOperator(RollOperator.SUB, xs[0], xs[1])
-    )
-    op_mul = value << "*" & expr > (
-        lambda xs: BinaryOperator(RollOperator.MUL, xs[0], xs[1])
-    )
-    op_div = value << "/" & expr > (
-        lambda xs: BinaryOperator(RollOperator.DIV, xs[0], xs[1])
-    )
-    op_pow = value << "^" & expr > (
-        lambda xs: BinaryOperator(RollOperator.POW, xs[0], xs[1])
-    )
-
-    expr = op_add | op_sub | op_mul | op_div | op_pow | value
-    brackets = "(" >> expr << ")"
-
-    parse_all = expr
-
-
-class ExcessiveDiceRollsError(Exception):
-    """Raised when too many dice are rolled in a single command"""
+def run(message, display_name):
+    try:
+        if len(message) == 0:
+            message = "1d6"
+        message = message.strip()
+        logging.debug("\n==== Parsing ====")
+        program = parse_program(message)
+        logging.debug("\n==== Evaluation ====")
+        logging.debug(program)
+        values = program.reduce()
+        logging.debug("\n==== Output ====")
+        string_rep = program.string_rep
+        pairs_assignments = string_rep.assignments
+        pairs_expressions = [
+            (values[i], string_rep.expressions[i]) for i in range(len(values))
+        ]
+        out = SUCCESS_OUT.format(
+            ping=display_name,
+            # body="\n".join([f"> **{pair[0]}**" for pair in pairs])
+            body="\n".join(
+                [f"{pair[0]} = `{pair[1]}`" for pair in pairs_assignments]
+                + [f"**{pair[0]}** ⟵ `{pair[1]}`" for pair in pairs_expressions]
+            ),
+        )
+        if len(out) > MAX_ROLLS:
+            raise OutputTooLargeError
+    except rollerr.WarningError as e:
+        out = WARNING_OUT.format(
+            ping=display_name, error=e.__class__.__name__, body=f"_{e.out}_"
+        )
+    except ParseError as e:
+        out = FAILURE_OUT.format(
+            ping=display_name, error=e.__class__.__name__, body=f"```{e}```"
+        )
+    except rollerr.RunTimeError as e:
+        out = FAILURE_OUT.format(
+            ping=display_name, error=e.__class__.__name__, body=f"```{e}```"
+        )
+    except rollerr.TypeError as e:
+        out = TYPE_ERROR_OUT.format(
+            ping=display_name, error=e.__class__.__name__, body=f"```{e}```"
+        )

```

Comment:
This block will never be caught as the `RunTimeError` above it is a superclass. Move this block above `RunTimeError`.